File: src/inscore/model/IText.cpp
/*

  INScore Project

  Copyright (C) 2009,2010  Grame

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

  Grame Research Laboratory, 11 cours de Verdun Gensoul, 69002 Lyon - France
  research@grame.fr

*/

#include <sstream>
#include <map>

#include "IAppl.h"
#include "IText.h"
#include "IMessage.h"
#include "Tools.h"
#include "Updater.h"
#include "VObjectView.h"

using namespace std;

namespace inscore
{

//--------------------------------------------------------------------------
const string IText::kTextType("txt");

// Font style and font weight constants
const string IText::kStyleNormal("normal");
const string IText::kStyleItalic("italic");
const string IText::kStyleOblique("oblique");
const string IText::kWeightNormal("normal");
const string IText::kWeightLight("light");
const string IText::kWeightDemiBold("demibold");
const string IText::kWeightBold("bold");
const string IText::kWeightBlack("black");

//--------------------------------------------------------------------------
IText::IText( const std::string& name, IObject * parent ) : IGraphicBasedObject(name, parent)
{ 
	fTypeString = kTextType;
	setColor( IColor(0,0,0,255) );

	// Default font parameters
#if 0
	fFontSize = 13;				// this was the previous setting, before a default font has been embedded in the appl ressources
	fFontFamily = "Arial";
#else
	fFontSize = 14;
	fFontFamily = IAppl::defaultFontName();
#endif
	fFontWeight = kWeightNormal;
	fFontStyle = kStyleNormal;
	fGetMsgHandlerMap[""] = TGetParamMsgHandler<string>::create(fText);

	// Method to manage font
	fGetMsgHandlerMap[kfontSize_GetSetMethod]	= TGetParamMethodHandler<IText, int (IText::*)() const>::create(this, &IText::getFontSize);
	fMsgHandlerMap[kfontSize_GetSetMethod]		= TSetMethodMsgHandler<IText, int>::create(this, &IText::setFontSize);

	fGetMsgHandlerMap[kfontFamily_GetSetMethod]	= TGetParamMethodHandler<IText, string (IText::*)() const>::create(this, &IText::getFontFamily);
	fMsgHandlerMap[kfontFamily_GetSetMethod]	= TSetMethodMsgHandler<IText, string>::create(this, &IText::setFontFamily);

	fGetMsgHandlerMap[kfontStyle_GetSetMethod]	= TGetParamMethodHandler<IText, string (IText::*)() const>::create(this, &IText::getFontStyle);
	fMsgHandlerMap[kfontStyle_GetSetMethod]		= TMethodMsgHandler<IText>::create(this, &IText::setFontStyle);

	fGetMsgHandlerMap[kfontWeight_GetSetMethod]	= TGetParamMethodHandler<IText, string (IText::*)() const>::create(this, &IText::getFontWeight);
	fMsgHandlerMap[kfontWeight_GetSetMethod]	= TMethodMsgHandler<IText>::create(this, &IText::setFontWeight);

	fMsgHandlerMap[kwrite_SetMethod]			= TMethodMsgHandler<IText, MsgHandler::msgStatus (IText::*)(const IMessage*)>::create(this, &IText::writeMsg);

#ifdef EMCC
	setPending();
#endif
}

//--------------------------------------------------------------------------
void IText::positionAble() {
	IObject::positionAble();
	// disable width and height methods for text. Use fontSize instead.
	fMsgHandlerMap[kwidth_GetSetMethod]		= nullptr;
	fMsgHandlerMap[kheight_GetSetMethod]	= nullptr;
	fGetMsgHandlerMap[kwidth_GetSetMethod]	= nullptr;
	fGetMsgHandlerMap[kheight_GetSetMethod]	= nullptr;
}

//--------------------------------------------------------------------------
void IText::accept (Updater* u)
{
	u->updateTo (SIText(this));
}

//--------------------------------------------------------------------------
void IText::print (ostream& out) const
{
	IObject::print (out);
	out << "  \"" << getText() << "\"";
}

//--------------------------------------------------------------------------
// message handlers
//--------------------------------------------------------------------------
MsgHandler::msgStatus IText::set( const IMessage* msg )
{ 
	MsgHandler::msgStatus status = IObject::set(msg);
	if (status & (MsgHandler::kProcessed + MsgHandler::kProcessedNoChange)) return status; 

	int n = msg->size();
	if (n >= 2) {
		stringstream stream;
		const char * sep = "";
		for (int i = 1; i < n ; i++) {
			string text; int itext; float ftext;
			stream << sep;
			if (msg->param(i, itext))			stream << itext;
			else if (msg->param(i, ftext))		stream << Tools::ensurefloat(ftext);
			else if (msg->param(i, text))		stream << text;
			sep = " ";
		}
		setText( stream.str() );
		newData(true);
		status = MsgHandler::kProcessed;
		getView()->initView (this);
	}
	else status = MsgHandler::kBadParameters;
	return status;
}


//--------------------------------------------------------------------------
void IText::setFontSize(int fontSize)
{
	fFontSize = fontSize;
	getView()->initView (this);
}

//--------------------------------------------------------------------------
MsgHandler::msgStatus IText::writeMsg (const IMessage* msg)
{
	if ((msg->size() >= 1)) {
		stringstream sstr;
		
		for (int i=0; i < msg->size(); i++) {
			sstr << " ";
			msg->print (sstr, i, 1, " ");
		}
		setText( fText + sstr.str() );
		newData(true);
		getView()->initView (this);
		return MsgHandler::kProcessed;
	}
	return MsgHandler::kBadParameters;
}

//--------------------------------------------------------------------------
MsgHandler::msgStatus IText::setFontWeight(const IMessage *msg)
{
	int n = msg->size();
	if (n == 1) {
		string fontWeight;
		if(msg->param(0, fontWeight)) {
			if(fontWeight == IText::kWeightNormal || fontWeight == IText::kWeightLight
					|| fontWeight == IText::kWeightDemiBold || fontWeight == IText::kWeightBold || fontWeight == IText::kWeightBlack) {
				fFontWeight = fontWeight;
				return MsgHandler::kProcessed;
			}
		}
	}
	return MsgHandler::kBadParameters;
}

//--------------------------------------------------------------------------
MsgHandler::msgStatus IText::setFontStyle(const IMessage *msg)
{
	int n = msg->size();
	if (n == 1) {
		string fontStyle;
		if(msg->param(0, fontStyle)) {
			if(fontStyle == IText::kStyleNormal || fontStyle == IText::kStyleItalic	|| fontStyle == IText::kStyleOblique) {
				fFontStyle = fontStyle;
				return MsgHandler::kProcessed;
			}
		}
	}
	return MsgHandler::kBadParameters;
}

}

File: src/Js/src/model/IText.ts
///<reference path="IObject.ts"/>

class IText extends IObject {
    protected fText: string;
	protected fFontSize: number;
	protected fFontFamily: string;
	protected fFontWeight: string;
	protected fFontStyle: string;
	protected fFontModified: boolean
    
    protected kTextType: string;

    // Constants for font style and font weight.
	static kStyleNormal = "normal" ;
	static kStyleItalic = "italic" ;
	static kStyleOblique = "oblique" ;
	static kWeightNormal = "normal" ;
	static kWeightLight = "light" ;
	static kWeightDemiBold = "demibold" ;
	static kWeightBold = "bold" ;
	static kWeightBlack = "black" ;
        
    constructor(name: string, parent: IObject) {
        super(name, parent);
        this.kTextType = 'txt';
        this.fTypeString = this.kTextType;
        
        //Default Parameters
        this.fColor.setRGB([0, 0, 0]);
        this.fColor.setA(255);
        this.fFontSize = 14;
        this.fFontFamily = "Arial"
        this.fFontWeight = IText.kWeightNormal;
	    this.fFontStyle = IText.kStyleNormal;
	    this.fFontModified = false;

     	this.fGetMsgHandlerMap[""] = new TGetMsgHandlerText(this._getText());
        super.setHandlers();
        
        // Method to manage font
        this.fGetMsgHandlerMap[kfontSize_GetSetMethod]      = new TGetMsgHandlerNum(this._getFontSize());
        this.fMsgHandlerMap[kfontSize_GetSetMethod] 	    = new TMsgHandlerNum(this._setFontSize());

        this.fGetMsgHandlerMap[kfontFamily_GetSetMethod]	= new TGetMsgHandlerText(this._getFontFamily());
        this.fMsgHandlerMap[kfontFamily_GetSetMethod]	    = new TMsgHandlerText(this._setFontFamily());

        this.fGetMsgHandlerMap[kfontStyle_GetSetMethod]	    = new TGetMsgHandlerText(this._getFontStyle());
        this.fMsgHandlerMap[kfontStyle_GetSetMethod]	    = new TMethodHandler(this._setFontStyle());

        this.fGetMsgHandlerMap[kfontWeight_GetSetMethod]	= new TGetMsgHandlerText(this._getFontWeight());
        this.fMsgHandlerMap[kfontWeight_GetSetMethod]	    = new TMethodHandler(this._setFontWeight());
    }    

// GETS / SETS VALUES 
//--------------------------------------------------------------    
    fontModified(): boolean         { return this.fFontModified; }
    getText(): string               { return this.fText; }
    setText(txt: string): void      { this.fText = txt; }
    
    getFontSize(): number           { return this.fFontSize; }
    setFontSize(s: number): void    { this.fFontSize = s; this.fFontModified = true;}
    
    getFontFamily(): string         { return this.fFontFamily; }
    setFontFamily(f: string): void  { this.fFontFamily = f; this.fFontModified = true; }
    
    getFontStyle(): string          { return this.fFontStyle; } 
    setFontStyle(msg: IMessage): msgStatus {
        let n = msg.size();
        if (n == 2) {
            let fontStyle = msg.paramStr(1);
            if(fontStyle.correct) {
                if(fontStyle.value == IText.kStyleNormal || fontStyle.value == IText.kStyleItalic	|| fontStyle.value == IText.kStyleOblique) {
                    this.fFontStyle = fontStyle.value;
                    this.fFontModified = true;
                    return msgStatus.kProcessed;
                }
            }
        }
        return msgStatus.kBadParameters;
    }       
    
    getFontWeight(): string         { return this.fFontWeight; }
    setFontWeight(msg: IMessage): msgStatus {
        let n = msg.size();
        if (n == 2) {
            let fontWeight = msg.paramStr(1);
            if(fontWeight.correct) {
                if(fontWeight.value == IText.kWeightNormal || fontWeight.value == IText.kWeightLight
                        || fontWeight.value == IText.kWeightDemiBold || fontWeight.value == IText.kWeightBold || fontWeight.value == IText.kWeightBlack) {
                    this.fFontWeight = fontWeight.value;
                    this.fFontModified = true;
                    return msgStatus.kProcessed;
                }
            }
        }
        return msgStatus.kBadParameters;
    }
    
// GETS / SETS VALUES CLOSURES
//--------------------------------------------------------------    
    _getText(): GetStringMethod         { return () => this.fText; }
    
    _getFontSize() : GetNumMethod       { return () => this.fFontSize; }
    _setFontSize() : SetNumMethod       { return (n) => this.setFontSize(n); }
    
    _getFontFamily() : GetStringMethod  { return () => this.fFontFamily; }
    _setFontFamily() : SetStringMethod  { return (s) => this.setFontFamily(s); }
    
    _getFontStyle() : GetStringMethod   { return () => this.fFontStyle; }
    _setFontStyle() : SetMsgMethod   	{ return (m) => this.setFontStyle(m); }

    _getFontWeight() : GetStringMethod  { return () => this.fFontWeight; }
    _setFontWeight() : SetMsgMethod  	{ return (m) => this.setFontWeight(m); }

// SET HANDLER
//--------------------------------------------------------------    
   set(msg: IMessage): msgStatus { 
        let status = super.set(msg);
        if (status & (msgStatus.kProcessed + msgStatus.kProcessedNoChange)) return status; 

        let n = msg.size();
        if (n > 2) {
            let str = "";
            let sep = "";
            for (let i = 2 ; i < n ; i++) {
                str += sep;
            	str += msg.param(i);
            	sep = " ";
            }
            this.setText(str);
            this.newData(true);
            status = msgStatus.kProcessed;
        }
        else status = msgStatus.kBadParameters;
        return status;
    }

// GETSET METHOD
//--------------------------------------------------------------    
    getSet(): IMessage	{ 
    	let a: Array<any> = [kset_SetMethod, this.kTextType];
    	return new IMessage(this.getOSCAddress(), a.concat (this.fText) ); 
    }
}